docs: improve databases documentation (#7732)
aquasecurity/trivy: docs/docs/configuration/db.md

# Trivy Databases

When you install Trivy, the installed artifact contains the scanner engine but is lacking relevant security information needed to make security detections and recommendations. These so called "databases" are fetched and maintained by Trivy automatically as needed, so normally you shouldn't notice or worry about them.   
This document elaborates on the database management mechanism and it's configuration options.

Trivy relies on the following databases:

DB | Artifact name | Contents | Purpose
--- | --- | --- | ---
Vulnerabilities DB | `trivy-db` | CVE information collected from various feeds | used only for [vulnerability scanning](../scanner/vulnerability.md)
Java DB | `trivy-java-db` | Index of Java artifacts and their hash digest | used to identify Java artifacts only in [JAR scanning](../coverage/language/java.md)

> Misconfiguration DB | `trivy-chekcs` | Logic of misconfiguration checks | used only in [misconfiguration/IaC scanning](../scanner/misconfiguration/check/builtin.md)

Author: I wanted to differentiate "databases" from "external resources". db is maintained by the trivy team and is fundemental to trivy operation. other external resources might be used by trivy like Maven central, which are not considered "databases". VEx repositories currently fall into the latter category I think

!!! note
    This is not an exhaustive list of Trivy's external connectivity requirements. There are additional external resources that might be required by specific Trivy features. To learn about external connectivity requirements, refer to the [Network connectivity document](../advanced/air-gap.md).

## Locations

Trivy's databases are published to the following locations:

| Registry | Image Address | Link
| --- | --- | ---
| GHCR | `ghcr.io/aquasecurity/trivy-db` | <https://ghcr.io/aquasecurity/trivy-db>
| | `ghcr.io/aquasecurity/trivy-java-db` | <https://ghcr.io/aquasecurity/trivy-java-db>
| | `ghcr.io/aquasecurity/trivy-checks` | <https://ghcr.io/aquasecurity/trivy-checks>
| Docker Hub | `aquasec/trivy-db` | <https://hub.docker.com/r/aquasec/trivy-db>
| | `aquasec/trivy-java-db` | <https://hub.docker.com/r/aquasec/trivy-java-db>
| AWS ECR | `public.ecr.aws/aquasecurity/trivy-db` | <https://gallery.ecr.aws/aquasecurity/trivy-db>
| | `public.ecr.aws/aquasecurity/trivy-java-db` | <https://gallery.ecr.aws/aquasecurity/trivy-java-db>

## Pulling and fallback

> Trivy will attempt to pull images from the following registries in the order specified. In case of a transient errors (e.g. status 429 or 5xx), Trivy will fall back to alternative registries.

Author: the section was rewritten lmk if it's still unclear

1. `ghcr.io/aquasecurity`

You can specify additional alternative repositories as explained in the [configuring database locations section](#locations).

Note that the Checks Database registry location option does not support fallback through multiple options. This is because in case of a failure pulling the trivy-checks DB, Trivy will use the embedded checks as a fallback.

## DB Management Configuration

### Database Locations

You can configure Trivy to download databases from alternative locations by using the flags:

- `--db-repository`
- `--java-db-repository`
- `--checks-bundle-repository`

> 

Author: see response to other similar comment

The value should be an image address in a container registry.

For example:

```
trivy image --db-repository registry.gitlab.com/gitlab-org/security-products/dependencies/trivy-db alpine
```

The flags accepts multiple values, which can be used to specify multiple alternative repository locations. In case of a transient errors (e.g. status 429 or 5xx), Trivy will fall back to alternative registries in the order specified.

For example:

```

> trivy image --db-repository my.registry.local/trivy-db --db-repository registry.gitlab.com/gitlab-org/security-products/dependencies/trivy-db alpine

Author: I think the example is self explanatory, especially since it the preceding example is exactly the same jsut with one flag, and is supposedly clear. this example just shows how specify multiple values

```

!!! note 
    Setting the repository location flags override the default values which include the official db locations. In case you want to preserve the default locations, you should include them in the list the you set as repository locations.

!!!note
    When pulling `trivy-db` or `trivy-java-db`, if image tag is not specified, Trivy defaults to the db schema number instead of the `latest` tag.

### Skip updates

You can configure Trivy to not attempt to download database at all using the flags:

- `--skip-db-update`
- `--skip-java-db-update`
- `--skip-check-update`

For example:

```
trivy image --skip-db-update --skip-java-db-update --offline-scan --skip-check-update myimage
```

### Only update

> You can ask `Trivy` to update the database without performing a scan using the flags:

Author: yes. updated

- `--download-db-only`
- `--download-java-db-only`

> 

Author: added. cc: @simar7

For example:

```
trivy image --download-db-only
```

### Remove Databases

`trivy clean` command removes caches and databases.
You can select which cache component to remove:

option | description
--- | ---
`-a`/`--all` | remove all caches
`--checks-bundle` | remove checks bundle
`--java-db` | remove Java database
`--scan-cache` | remove scan cache (container and VM image analysis results)

> `--vuln-db` | remove vulnerability database

Author: agree. I also mention this in several previous issues (that we need to standardize the names of databases) cc @simar7 @knqyf263 . For now that state is accurate

# Trivy Databases

> When you install Trivy, the installed artifact contains the scanner engine but is lacking relevant security information needed to make security detections and recommendations. These so called "databases" are fetched and maintained by Trivy automatically as needed, so normally you shouldn't notice or worry about them. However, some situations might require your attention to Trivy's network connectivity. This section elaborates on the database management mechanism and it's configuration options.

Reviewer: `scanner engine but is lacking relevant security information needed to make security detections and recommendations. ` we repeat this in the air gap doc as well. Is this intentional?

Author: I thought it's worth reintroducing the motivation since someone might be reading one doc and not the other. I you think it's unnesessary I can remove the intro from the air gap doc.

| Registry | Image Address | Link
| --- | --- | ---
| GHCR | `ghcr.io/aquasecurity/trivy-db:2` | <https://ghcr.io/aquasecurity/trivy-db>
| | `ghcr.io/aquasecurity/trivy-java-db` | <https://ghcr.io/aquasecurity/trivy-java-db>
| | `aquaghcr.io/aquasecurity/ecurity/trivy-checks` | <https://ghcr.io/aquasecurity/trivy-checks>
| Docker Hub | `aquasec/trivy-db:2` | <https://hub.docker.com/r/aquasec/trivy-db>
| | `aquasec/trivy-java-db:1` | <https://hub.docker.com/r/aquasec/trivy-java-db>
| AWS ECR | `public.ecr.aws/aquasecurity/trivy-db:2` | <https://gallery.ecr.aws/aquasecurity/trivy-db>
| | `public.ecr.aws/aquasecurity/trivy-java-db:1` | <https://gallery.ecr.aws/aquasecurity/trivy-java-db>

 VEX Hub is fetched from VEX Hub GitHub Repository directly: <https://github.com/aquasecurity/vexhub>.

### Automatic fallback

Trivy will attempt to pull images from the official registries in the order specified. In case of failure of pulling a database, Trivy will fall back to the next alternative registry in the order specified.  
You can specify additional alternative repositories as explained in the [configuring database locations section](#locations).

> The Checks Database registry location option does not support fallback through multiple options.

Reviewer: ```suggestion
The Checks Database registry location option does not support fallback through multiple options. This is because in case of a failure pulling the checks-db, Trivy will use the embedded checks as a fallback.
```

## Connectivity requirements

> | database | location | protocol | hosts
> | --- | --- | --- | ---
> | <ul><li>`trivy-db`</li><li>`trivy-java-db`</li><li>`checks-db`</li></ul> | GHCR | [OCI Distribution](https://github.com/opencontainers/distribution-spec) over HTTPS | <ul><li>`ghcr.io`</li><li>`pkg-containers.githubusercontent.com`</li></ul>
> | `vexhub`| GitHub | Git over HTTPS | <ul><li>`api.github.com`</li><li>`codeload.github.com`</li></ul>
> 

Reviewer: fixing table borders

```suggestion
| database                                                         | location | protocol                                                                           | hosts                                                                      |
|------------------------------------------------------------------|----------|------------------------------------------------------------------------------------|----------------------------------------------------------------------------|
| `trivy-db`</li><li>`trivy-java-db`</li><li>`checks-db`</li></ul> | GHCR     | [OCI Distribution](https://github.com/opencontainers/distribution-spec) over HTTPS | <ul><li>`ghcr.io`</li><li>`pkg-containers.githubusercontent.com`</li></ul> |
| `vexhub`                                                         | GitHub   | Git over HTTPS                                                                     | <ul><li>`api.github.com`</li><li>`codeload.github.com`</li></ul>           |

```

Reviewer: Should we separate out the three DBs into their own rows?

Author: actually I don't like adding extra spaces and dashes in markdown, it makes editing very annoying

For more information about GitHub connectivity (including specific IP addresses), please refer to [GitHub's connectivity troubleshooting guide](https://docs.github.com/en/get-started/using-github/troubleshooting-connectivity-problems).

### Rate limiting

> Trivy is an open source project that relies on public free infrastructure. In case of extreme load, you may encounter rate limiting when Trivy attempts to update its databases. If you are facing rate-limiting issues:

Reviewer: formatting
```suggestion
Trivy is an open source project that relies on public free infrastructure. In case of extreme load, you may encounter rate limiting when Trivy attempts to update its databases. If you are facing rate-limiting issues:

```

3. Consider using a commercial service that provides the full [Aqua platform](https://www.aquasec.com/products/software-supply-chain-security/), which includes a stable and reliable scanning service.

## DB Management Configuration

### Database Locations

You can configure Trivy to download databases from alternative locations by using the flags:

- `--db-repository`
- `--java-db-repository`
- `--checks-bundle-repository`

The value should be an image address in a container registry.

For example:

```
trivy image --db-repository registry.gitlab.com/gitlab-org/security-products/dependencies/trivy-db alpine
```

The `--db-repository` flag accepts multiple values, which can be used to specify multiple alternative repository locations. In case of failure, Trivy will fall back to alternative registries in the order specified. An attempt to download from the next repository is only made if a temporary error is received (e.g. status 429 or 5xx).

For example:

```
trivy image --db-repository my.registry.local/trivy-db,registry.gitlab.com/gitlab-org/security-products/dependencies/trivy-db alpine
```

!!! note 

>     setting the repository location flags override the default values which include the official db locations. In case you want to preserve the default locations, you should include them in the list the you set as repository locations.

Reviewer: ```suggestion
    Setting the repository location flags override the default values which include the official db locations. In case you want to preserve the default locations, you should include them in the list the you set as repository locations.
```

!!!note

>     When pulling `trivy-db` or `trivy-java-db`, if image tag is not specified, Trivy defaults to the db schema number instead of the `latest` tag. Currently, for `trivy-db` the default tag is `2`, for `trivy-java-db` the default tag is `1`, for `trivy-checks` the default tag is `0`.

Reviewer: I would remove specific versions from this because they will get out of date easily. Already the check-db is at default tag of 1.

```suggestion
    When pulling `trivy-db` or `trivy-java-db`, if image tag is not specified, Trivy defaults to the db schema number instead of the `latest` tag.
```

```

### Remove Databases

`trivy clean` command removes caches and databases.
You can select which cache component to remove:

option | description
--- | ---
`-a`/`--all` | remove all caches
`--checks-bundle` | remove checks bundle
`--java-db` | remove Java database
`--scan-cache` | remove scan cache (container and VM image analysis results)
`--vex-repo` | remove VEX repositories
`--vuln-db` | remove vulnerability database

Example:

```
$ trivy clean --vuln-db --java-db
2024-06-24T11:42:31+06:00       INFO    Removing vulnerability database...
2024-06-24T11:42:31+06:00       INFO    Removing Java database...
```

## Self-Hosting

You can host all of Trivy's databases on your own local environment (to prevent external connectivity). For more information, refer to the [Air-Gapped Environments and Self-Hosting](../advanced/air-gap.md) document.

When serving, proxying, or manipulating Trivy's databases, note that the media type of the OCI layer is not a standard container image type.

DB | Media Type | Reference
--- | --- | ---
`trivy-db` | `application/vnd.aquasec.trivy.db.layer.v1.tar+gzip` | <https://github.com/aquasecurity/trivy-db/pkgs/container/trivy-db>

> `trivy-java-db` | `application/vnd.aquasec.trivy.javadb.layer.v1.tar+gzip` | https://github.com/aquasecurity/trivy-java-db/pkgs/container/trivy-java-db

Reviewer: ```suggestion
`trivy-java-db` | `application/vnd.aquasec.trivy.javadb.layer.v1.tar+gzip` | <https://github.com/aquasecurity/trivy-java-db/pkgs/container/trivy-java-db>
```

Reviewer: You can also add the checks db from https://github.com/aquasecurity/trivy-checks/pkgs/container/trivy-checks 

`"mediaType": "application/vnd.oci.image.manifest.v1+json"`

# Trivy Databases

When you install Trivy, the installed artifact contains the scanner engine but is lacking relevant security information needed to make security detections and recommendations. These so called "databases" are fetched and maintained by Trivy automatically as needed, so normally you shouldn't notice or worry about them. However, some situations might require your attention to Trivy's network connectivity. This section elaborates on the database management mechanism and it's configuration options.

Trivy relies on the following databases:

DB | Artifact name | Contents | Purpose
--- | --- | --- | ---
Vulnerabilities DB | `trivy-db` | CVE information collected from various feeds | used only for [vulnerability scanning](../scanner/vulnerability.md)

> Java DB | `trivy-java-db` | Index of Java artifacts and their hash digest | used to identify Java artifacts only in [Java vulnerability scanning](../coverage/language/java.md)

Reviewer 2: trivy-java-db is not used for POM scanning, and also needed for SBOM.

```suggestion
Java DB | `trivy-java-db` | Index of Java artifacts and their hash digest | used to identify Java artifacts only in [JAR scanning](../coverage/language/java.md)
```

Misconfiguration DB | `trivy-chekcs` | Logic of misconfiguration checks | used only in [misconfiguration/IaC scanning](../scanner/misconfiguration/check/builtin.md)

> VEX Hub | `vex-hub` | VEX statements | Used only in [VEX Hub is enabled](../supply-chain/vex/repo.md) in vulnerability scanning.

Reviewer 2: ```suggestion
VEX Hub | `vex-hub` | VEX statements | Used only in [VEX Repository is enabled](../supply-chain/vex/repo.md) in vulnerability scanning.
```

## External Services

In addition to the above, some specific scanning features might connect to external services, and have different connectivity requirements and configuration options. This document discusses only Trivy's own databases, but for your convenience here are use cases where external services are involved:

- [Java vulnerability scanning](../coverage/language/java.md).

## Locations

Following are official locations of Trivy databases:

> | Registry | Image Address | Link

Reviewer 2: Shouldn't we mention which one is the default?

Author: I thought that all official repos should be used by default, in the order specified. but I realize this is not the case.

| --- | --- | ---

> | GHCR | `ghcr.io/aquasecurity/trivy-db` | <https://ghcr.io/aquasecurity/trivy-db>

Reviewer 2: I think it's more straightforward to aggregate them by artifacts so users can understand which address they can use for secondary.

- trivy-db
  - ghcr.io/aquasecurity/trivy-db (default)
  - aquasec/trivy-db
  - public.ecr.aws/aquasecurity/trivy-db
- trivy-java-db
  - ghcr.io/aquasecurity/trivy-java-db (default)
  - ...
--

<br class="Apple-interchange-newline">

Author: I'm fine with making the change, but just thinking users might have a mirror preference based on their infrastructure, for example an AWS customer would probably prefer ECR for all of the images.

> | GHCR | `ghcr.io/aquasecurity/trivy-db:2` | <https://ghcr.io/aquasecurity/trivy-db>

Reviewer 3: Why are some images with tags and some without?

| | `ghcr.io/aquasecurity/trivy-java-db` | <https://ghcr.io/aquasecurity/trivy-java-db>

> | | `aquaghcr.io/aquasecurity/ecurity/trivy-checks` | <https://ghcr.io/aquasecurity/trivy-checks>

Reviewer 3: Should be ghcr.io/aquasecurity/trivy-checks here?

| Docker Hub | `aquasec/trivy-db` | <https://hub.docker.com/r/aquasec/trivy-db>
| | `aquasec/trivy-java-db` | <https://hub.docker.com/r/aquasec/trivy-java-db>
| AWS ECR | `public.ecr.aws/aquasecurity/trivy-db` | <https://gallery.ecr.aws/aquasecurity/trivy-db>
| | `public.ecr.aws/aquasecurity/trivy-java-db` | <https://gallery.ecr.aws/aquasecurity/trivy-java-db>

 VEX Hub is fetched from VEX Hub GitHub Repository directly: <https://github.com/aquasecurity/vexhub>.

### Automatic fallback

> Trivy will attempt to pull images from the official registries in the order specified. In case of a transient errors (e.g. status 429 or 5xx), Trivy will fall back to alternative registries in the order specified.  

Reviewer 2: It's repeated.

```suggestion
Trivy will attempt to pull images from the official registries in the order specified. In case of a transient errors (e.g. status 429 or 5xx), Trivy will fall back to alternative registries.  
```

> Trivy will attempt to pull images from the official registries in the order specified. In case of failure of pulling a database, Trivy will fall back to the next alternative registry in the order specified.  

Reviewer 3: The database will be loaded from the next repository only in case of a temporary error. (429 status and some 5xx) https://github.com/aquasecurity/trivy/pull/7605#discussion_r1780802178

Author: it's mentioned here generally and more detailed later in the flags documentation. but ok, added here too

You can specify additional alternative repositories as explained in the [configuring database locations section](#locations).

> The Checks Database registry location option does not support fallback through multiple options. This is because in case of a failure pulling the trivy-chekcs DB, Trivy will use the embedded checks as a fallback.

Reviewer 2: We call it "checks bundle" in all other places. Do we want to rename it officially?

Author: I'm fine with it.

Reviewer 2: @simar7 I think you named it "checks bundle." What do you think?

Reviewer: Yeah checks bundle is good with me.

Reviewer 2: @itaysk is calling it "checks database" in this PR. We should be consistent in what we call it within the project. Either is fine for me.

- Checks Bundle
- Checks Database

Note we already have a flag `--checks-bundle-repository` and need to rename it to `--checks-db-repository` if we'll go with checks database.

```
      --checks-bundle-repository string   OCI registry URL to retrieve checks bundle from (default "ghcr.io/aquasecurity/trivy-checks:1")
```

Reviewer: I'm also fine with either.

## Connectivity requirements

| database | location | protocol | hosts
| --- | --- | --- | ---
| <ul><li>`trivy-db`</li><li>`trivy-java-db`</li><li>`trivy-chekcs`</li></ul> | GHCR | [OCI Distribution](https://github.com/opencontainers/distribution-spec) over HTTPS | <ul><li>`ghcr.io`</li><li>`pkg-containers.githubusercontent.com`</li></ul>

> | `vexhub`| GitHub | Git over HTTPS | <ul><li>`api.github.com`</li><li>`codeload.github.com`</li></ul>

Reviewer 2: The git protocol is not used for VEX repositories.

```suggestion
| `vexhub`| GitHub | HTTPS | <ul><li>`api.github.com`</li><li>`codeload.github.com`</li></ul>
```

For more information about GitHub connectivity (including specific IP addresses), please refer to [GitHub's connectivity troubleshooting guide](https://docs.github.com/en/get-started/using-github/troubleshooting-connectivity-problems).

### Rate limiting

Trivy is an open source project that relies on public free infrastructure. In case of extreme load, you may encounter rate limiting when Trivy attempts to update its databases. If you are facing rate-limiting issues:

> 1. Consider self-hosting the databases, or implementing a proxy-cache in your organization.

Reviewer 2: Adding a  link to the page about self-hosting?

2. Look into specific registry rate-limiting policies which might provide solution for your use-case (for example by authenticating with the registry).
3. Consider using a commercial service that provides the full [Aqua platform](https://www.aquasec.com/products/software-supply-chain-security/), which includes a stable and reliable scanning service.

## DB Management Configuration

### Database Locations

You can configure Trivy to download databases from alternative locations by using the flags:

- `--db-repository`
- `--java-db-repository`
- `--checks-bundle-repository`

The value should be an image address in a container registry.

For example:

```
trivy image --db-repository registry.gitlab.com/gitlab-org/security-products/dependencies/trivy-db alpine
```

> The flags accepts multiple values, which can be used to specify multiple alternative repository locations. In case of a transient errors (e.g. status 429 or 5xx), Trivy will fall back to alternative registries in the order specified.

Reviewer 2: ```suggestion
The flags accepts multiple values, which can be used to specify multiple alternative repository locations. See [Automatic fallback](#automatic-fallback) for details.
```

Author: I thought that the authoritative section is this one, which describes all the db flags, and the db locations section on top should point here, not the other way around.

For example:

```

> trivy image --db-repository my.registry.local/trivy-db,registry.gitlab.com/gitlab-org/security-products/dependencies/trivy-db alpine

Reviewer 2: A comma also works, but the repeated flags may be easier to understand the intention.

```suggestion
trivy image --db-repository my.registry.local/trivy-db --db-repository registry.gitlab.com/gitlab-org/security-products/dependencies/trivy-db alpine
```
